### ISSUE TITLE
buck2 rebuilds all sources in a library when new source is added.

### ISSUE THREAD

**Issue by zjturner** [facebook/buck2#376]
I tried this two separate ways: with a glob and with an explicit source file list.  In both cases, if you add a new file to a library buck2 fully recompiles all *other* objects in the library.

Example 1:

```
$ cat BUCK
cxx_library(
    name="test",
    visibility=["PUBLIC"],
    srcs=["foo.cpp", "bar.cpp"]
)

$ buck2 build //:test && buck2 log what-ran
<snip>
Cache hits: 0%. Commands: 3 (cached: 0, remote: 0, local: 3)
BUILD SUCCEEDED
Showing commands from: buck2 build //:test
build   <snip>  deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\foo.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\foo.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\bar.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\bar.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/lib.exe "/OUT:buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib.argsfile"

$ touch baz.cpp
$ echo "Edit the BUCK file to add baz.cpp to the srcs list"
$ cat BUCK
cxx_library(
    name="test",
    visibility=["PUBLIC"],
    srcs=["foo.cpp", "bar.cpp", "baz.cpp"]
)

$ buck2 build //:test && buck2 log what-ran
<snip>
Cache hits: 0%. Commands: 4 (cached: 0, remote: 0, local: 4)
BUILD SUCCEEDED
Showing commands from: buck2 build //:test
build   <snip>  deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\baz.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\baz.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\bar.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\bar.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\foo.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\foo.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/lib.exe "/OUT:buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib.argsfile"
```

Example 2:
```
$ cat BUCK
cxx_library(
    name="test",
    visibility=["PUBLIC"],
    srcs=glob(["*.cpp"])
)

$ buck2 build //:test && buck2 log what-ran
<snip>
Cache hits: 0%. Commands: 3 (cached: 0, remote: 0, local: 3)
BUILD SUCCEEDED
Showing commands from: buck2 build //:test
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\foo.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\foo.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\bar.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\bar.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/lib.exe "/OUT:buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib.argsfile"

$ touch baz.cpp && buck2 build //:test && buck2 log what-ran
$ buck2 build //:test && buck2 log what-ran
<snip>
Cache hits: 0%. Commands: 4 (cached: 0, remote: 0, local: 4)
BUILD SUCCEEDED
Showing commands from: buck2 build //:test
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\baz.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\baz.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\foo.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\foo.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/cl.exe "/Fobuck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\__objects__\\bar.cpp.obj" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\.cpp.argsfile" -c ".\\bar.cpp"
build   <snip> deps/msvc-hermetic/msvc/bin/Hostx64/x64/lib.exe "/OUT:buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib" "@buck-out\\v2\\gen\\root\\23b3e4bdecafe14a\\__test__\\libtest.lib.argsfile"
```

I know that in a remote execution environment or with a cache this matters less.  But even with a cache, this is not ideal from a bandwidth perspective, and bandwidth = money.  So performance aside, there is still considerable cost savings to be had by making it a little bit smarter here.

**Comment by zjturner** [comment]
This is another case (similar to #340) where `hash_all_commands` seems to fix this, but it seems like this should be supported even without hashing commands.

**Comment by ndmitchell** [comment]
I'm not sure why it should necessarily work without `hash_all_commands` - the actions after we run analysis are distinct fresh actions, so you need to link them back to the previous ones. You can do that by indexing the actions by a stable key and doing some kind of equality, or by hashing the command, and since the latter has value anyway that seems reasonable. @krallin - is there a reason `hash_all_commands` isn't the default?

I tried this benchmark internally and it worked, because we do have hash_all_commands turned on.

**Comment by krallin** [comment]
> @krallin - is there a reason hash_all_commands isn't the default?

Because it requires the deferred materializer. I guess if you were an end user you could just enable that + paranoid mode and you'd be happy.

Alternatively, we can add support for kicking off materialization immediately (in the background) in the deferred mateiralizer  and I think you could make it the default easily.

**Comment by krallin** [comment]
> but hashing commands is still work

I'd be very surprised if you were to ever see this (and I'll note checking mtimes is also work if you have to check a lot of mtimes — hashing commands is something Buck2 does without doing any I/O to disk).

**Comment by krallin** [comment]
TL;DR, if I were you, I would suggest actually giving this `.buckconfig` a shot:

```
[buck2]
materializations = deferred
sqlite_materializer_state = true
defer_write_actions = true
hash_all_commands = true
restarter = true
```

If you're using RE, the risk of enabling the deferred materialization is that if outputs expire in your RE backend, Buck2 might try to download them (with a fairly descriptive error message that invites you to restart your Buck2 daemon). Enabling the restarter should mitigate that as it'll just restart your Buck2 if you hit this conditon.

**Comment by krallin** [comment]
In fairness the caveats of deferred materialization is literally the only thing we document our buckconfigs!

https://www.internalfb.com/intern/staticdocs/buck2/docs/users/advanced/deferred_materialization/

**Comment by krallin** [comment]
Ah yes, here it is:

https://buck2.build/docs/users/advanced/deferred_materialization/
